wolfcrypt/src/port/psa/psa_hash.c

```diff
@@ -0,0 +1,300 @@
+        psa_hash_abort(ctx);
+        PSA_UNLOCK();
+        return WC_HW_E;
+    }
+
+    return 0;
+}
+
+static int wc_psa_hash_update(psa_hash_operation_t *ctx, const uint8_t *input,
+                              size_t input_length)
+{
+    psa_status_t s;
+
+    if (ctx == NULL || (input == NULL && input_length > 0))
+        return BAD_FUNC_ARG;
+
+    PSA_LOCK();
+    s = psa_hash_update(ctx, input, input_length);
+    PSA_UNLOCK();
+
+    if (s != PSA_SUCCESS) {
+        PSA_LOCK();
+        psa_hash_abort(ctx);
+        PSA_UNLOCK();
+        return WC_HW_E;
+    }
+
+    return 0;
+}
+
+static int wc_psa_hash_finish_setup(psa_hash_operation_t *ctx,
+                                    uint8_t *output, psa_algorithm_t alg)
+{
+    size_t hash_length;
+    psa_status_t s;
+
+    if (ctx == NULL || output == NULL)
+        return BAD_FUNC_ARG;
+
+    PSA_LOCK();
+    s = psa_hash_finish(ctx, output, PSA_HASH_LENGTH(alg), &hash_length);
+    PSA_UNLOCK();
+    if (s != PSA_SUCCESS) {
+        PSA_LOCK();
+        psa_hash_abort(ctx);
+        PSA_UNLOCK();
+        return WC_HW_E;
+    }
+
+    PSA_LOCK();
+    s = psa_hash_setup(ctx, alg);
+    PSA_UNLOCK();
+    if (s != PSA_SUCCESS) {
+        PSA_LOCK();
+        psa_hash_abort(ctx);
+        PSA_UNLOCK();
+        return WC_HW_E;
+    }
+
+    return 0;
+}
+
+static int wc_psa_hash_clone(const psa_hash_operation_t *src,
+                             psa_hash_operation_t *dst)
+{
+    psa_status_t s;
+
+    if (src == NULL || dst == NULL)
+        return BAD_FUNC_ARG;
+
+    PSA_LOCK();
+    psa_hash_abort(dst);
+    PSA_UNLOCK();
+
+    PSA_LOCK();
+    s = psa_hash_clone(src, dst);
+    PSA_UNLOCK();
+
+    if (s != PSA_SUCCESS)
+        return WC_HW_E;
+
+    return 0;
+}
+
+static int wc_psa_hash_abort(psa_hash_operation_t *ctx)
+{
+    psa_status_t s;
+
+    if (ctx == NULL)
+        return BAD_FUNC_ARG;
+
+    PSA_LOCK();
+    s = psa_hash_abort(ctx);
+    PSA_UNLOCK();
+    if (s != PSA_SUCCESS)
+        return WC_HW_E;
+
+    return 0;
+}
+
+static int wc_psa_get_hash(psa_hash_operation_t *ctx,
+                           uint8_t *out, psa_algorithm_t alg)
+{
+    WC_MAYBE_UNUSED size_t hash_length;
+    psa_hash_operation_t tmp;
+    psa_status_t s;
+
+    if (ctx == NULL || out == NULL)
+        return BAD_FUNC_ARG;
+
+    XMEMSET(&tmp, 0, sizeof(tmp));
+    PSA_LOCK();
+    s = psa_hash_clone(ctx, &tmp);
+    PSA_UNLOCK();
+    if (s != PSA_SUCCESS) {
+        psa_hash_abort(&tmp);
+        return WC_HW_E;
+    }
+
+    PSA_LOCK();
+    s = psa_hash_finish(&tmp, out, PSA_HASH_LENGTH(alg), &hash_length);
+    PSA_UNLOCK();
+    if (s != PSA_SUCCESS) {
+        PSA_LOCK();
+        psa_hash_abort(&tmp);
+        PSA_UNLOCK();
+        return WC_HW_E;
+    }
+
+    return 0;
+}
+#endif /* !defined(NO_SHA) || !defined(NO_SHA256)  || !defined(NO_SHA224)*/
+
+#if !defined(NO_SHA)
+
+int wc_InitSha_ex(wc_Sha *sha,
+                              void *heap WC_MAYBE_UNUSED,
```

Comment:
Worried the `WC_MAYBE_UNUSED` here isn't portable. I would prefer just adding the `(void)` for these in the function like we do elsewhere in the library.